File: src/main/api/catalog.ts
import { generateRefferer, generateSiteId } from './utils'

interface Config {
  baseUrl: string
  userAgent: string
  test?: boolean
}

// siteid : name ; 1 : manga ; 2 : slash ; 3 : ranobe ; 4: hentai ; 5: anime ;

export class CatalogManager {
  private config
  constructor(config: Config) {
    this.config = config
  }
  async searchMangaObjects(
    query: string,
    fields: string[] = ['rate_avg', 'rate', 'releaseDate'],
    siteIds: number[] = [1, 2, 3, 4]
  ): Promise<string> {
    const params = [
      `q=${encodeURIComponent(query)}`,
      ...fields.map((field) => `fields[]=${encodeURIComponent(field)}`),
      ...siteIds.map((siteId) => `site_id[]=${siteId}`)
    ].join('&')

    try {
      const response = await fetch(
        `https://${this.config.baseUrl}/api/manga?${params.toString()}`,
        {
          credentials: 'include',
          headers: {
            'User-Agent': this.config.userAgent,
            Accept: '*/*',
            'Accept-Language': 'ru-RU,en-US,en', //navigator.languages.join(','),
            'Site-Id': generateSiteId('manga'),
            'Content-Type': 'application/json',
            'Client-Time-Zone': Intl.DateTimeFormat().resolvedOptions().timeZone,
            'Sec-Fetch-Dest': 'empty',
            'Sec-Fetch-Mode': 'cors',
            'Sec-Fetch-Site': 'cross-site'
          },
          referrer: generateRefferer('manga'),
          method: 'GET',
          mode: 'cors'
        }
      )

      if (!response.ok) {
        return `{"http": "${response.status}"}`
      }

      const data = await response.json()
      return data
    } catch (error) {
      return `{"error": "${error}"}`
    }
  }

  async searchAnimeObjects(
    query: string,
    fields: string[] = ['rate_avg', 'rate', 'releaseDate']
  ): Promise<string> {
    const params = [
      `q=${encodeURIComponent(query)}`,
      ...fields.map((field) => `fields[]=${encodeURIComponent(field)}`)
    ].join('&')

    try {
      const response = await fetch(
        `https://${this.config.baseUrl}/api/anime?${params.toString()}`,
        {
          credentials: 'include',
          headers: {
            'User-Agent': this.config.userAgent,
            Accept: '*/*',
            'Accept-Language': 'ru-RU,en-US,en',
            'Content-Type': 'application/json',
            'Site-Id': generateSiteId('anime'),
            'Client-Time-Zone': Intl.DateTimeFormat().resolvedOptions().timeZone,
            'Sec-Fetch-Dest': 'empty',
            'Sec-Fetch-Mode': 'cors',
            'Sec-Fetch-Site': 'cross-site'
          },
          referrer: generateRefferer('anime'),
          method: 'GET',
          mode: 'cors'
        }
      )

      if (!response.ok) {
        return `{"http": "${response.status}"}`
      }

      const data = await response.json()
      return data
    } catch (error) {
      return `{"error": "${error}"}`
    }
  }

  async searchTeams(query: string): Promise<string> {
    const params = [`q=${encodeURIComponent(query)}`].join('&')
    try {
      const response = await fetch(
        `https://${this.config.baseUrl}/api/teams?${params.toString()}`,
        {
          credentials: 'include',
          headers: {
            'User-Agent': this.config.userAgent,
            Accept: '*/*',
            'Accept-Language': 'ru-RU,en-US,en',
            'Content-Type': 'application/json',
            'Site-Id': generateSiteId('anime'),
            'Client-Time-Zone': Intl.DateTimeFormat().resolvedOptions().timeZone,
            'Sec-Fetch-Dest': 'empty',
            'Sec-Fetch-Mode': 'cors',
            'Sec-Fetch-Site': 'cross-site'
          },
          referrer: generateRefferer('anime'),
          method: 'GET',
          mode: 'cors'
        }
      )

      if (!response.ok) {
        return `{"http": "${response.status}"}`
      }

      const data = await response.json()
      return data
    } catch (error) {
      return `{"error": "${error}"}`
    }
  }

  async searchCharacter(query: string): Promise<string> {
    const params = [`q=${encodeURIComponent(query)}`].join('&')
    try {
      const response = await fetch(
        `https://${this.config.baseUrl}/api/character?${params.toString()}`,
        {
          credentials: 'include',
          headers: {
            'User-Agent': this.config.userAgent,
            Accept: '*/*',
            'Accept-Language': 'ru-RU,en-US,en',
            'Content-Type': 'application/json',
            'Site-Id': generateSiteId('anime'),
            'Client-Time-Zone': Intl.DateTimeFormat().resolvedOptions().timeZone,
            'Sec-Fetch-Dest': 'empty',
            'Sec-Fetch-Mode': 'cors',
            'Sec-Fetch-Site': 'cross-site'
          },
          referrer: generateRefferer('anime'),
          method: 'GET',
          mode: 'cors'
        }
      )

      if (!response.ok) {
        return `{"http": "${response.status}"}`
      }

      const data = await response.json()
      return data
    } catch (error) {
      return `{"error": "${error}"}`
    }
  }

  async searchPeople(query: string): Promise<string> {
    const params = [`q=${encodeURIComponent(query)}`].join('&')
    try {
      const response = await fetch(
        `https://${this.config.baseUrl}/api/people?${params.toString()}`,
        {
          credentials: 'include',
          headers: {
            'User-Agent': this.config.userAgent,
            Accept: '*/*',
            'Accept-Language': 'ru-RU,en-US,en',
            'Content-Type': 'application/json',
            'Site-Id': generateSiteId('anime'),
            'Client-Time-Zone': Intl.DateTimeFormat().resolvedOptions().timeZone,
            'Sec-Fetch-Dest': 'empty',
            'Sec-Fetch-Mode': 'cors',
            'Sec-Fetch-Site': 'cross-site'
          },
          referrer: generateRefferer('anime'),
          method: 'GET',
          mode: 'cors'
        }
      )

      if (!response.ok) {
        return `{"http": "${response.status}"}`
      }

      const data = await response.json()
      return data
    } catch (error) {
      return `{"error": "${error}"}`
    }
  }

  async searchFranchise(query: string): Promise<string> {
    const params = [`q=${encodeURIComponent(query)}`].join('&')
    try {
      const response = await fetch(
        `https://${this.config.baseUrl}/api/franchise?${params.toString()}`,
        {
          credentials: 'include',
          headers: {
            'User-Agent': this.config.userAgent,
            Accept: '*/*',
            'Accept-Language': 'ru-RU,en-US,en',
            'Content-Type': 'application/json',
            'Site-Id': generateSiteId('anime'),
            'Client-Time-Zone': Intl.DateTimeFormat().resolvedOptions().timeZone,
            'Sec-Fetch-Dest': 'empty',
            'Sec-Fetch-Mode': 'cors',
            'Sec-Fetch-Site': 'cross-site'
          },
          referrer: generateRefferer('anime'),
          method: 'GET',
          mode: 'cors'
        }
      )

      if (!response.ok) {
        return `{"http": "${response.status}"}`
      }

      const data = await response.json()
      return data
    } catch (error) {
      return `{"error": "${error}"}`
    }
  }

  async searchPublisher(query: string): Promise<string> {
    const params = [`q=${encodeURIComponent(query)}`].join('&')
    try {
      const response = await fetch(
        `https://${this.config.baseUrl}/api/publisher?${params.toString()}`,
        {
          credentials: 'include',
          headers: {
            'User-Agent': this.config.userAgent,
            Accept: '*/*',
            'Accept-Language': 'ru-RU,en-US,en',
            'Content-Type': 'application/json',
            'Site-Id': generateSiteId('anime'),
            'Client-Time-Zone': Intl.DateTimeFormat().resolvedOptions().timeZone,
            'Sec-Fetch-Dest': 'empty',
            'Sec-Fetch-Mode': 'cors',
            'Sec-Fetch-Site': 'cross-site'
          },
          referrer: generateRefferer('anime'),
          method: 'GET',
          mode: 'cors'
        }
      )

      if (!response.ok) {
        return `{"http": "${response.status}"}`
      }

      const data = await response.json()
      return data
    } catch (error) {
      return `{"error": "${error}"}`
    }
  }

  async searchUser(
    query: string,
    sort_by: string = 'id',
    sort_type: string = 'asc'
  ): Promise<string> {
    const params = [
      `q=${encodeURIComponent(query)}`,
      `sort_by=${sort_by}`,
      `sort_type=${sort_type}`
    ].join('&')
    try {
      const response = await fetch(`https://${this.config.baseUrl}/api/user?${params.toString()}`, {
        credentials: 'include',
        headers: {
          'User-Agent': this.config.userAgent,
          Accept: '*/*',
          'Accept-Language': 'ru-RU,en-US,en',
          'Content-Type': 'application/json',
          'Site-Id': generateSiteId('anime'),
          'Client-Time-Zone': Intl.DateTimeFormat().resolvedOptions().timeZone,
          'Sec-Fetch-Dest': 'empty',
          'Sec-Fetch-Mode': 'cors',
          'Sec-Fetch-Site': 'cross-site'
        },
        referrer: generateRefferer('anime'),
        method: 'GET',
        mode: 'cors'
      })

      if (!response.ok) {
        return `{"http": "${response.status}"}`
      }

      const data = await response.json()
      return data
    } catch (error) {
      return `{"error": "${error}"}`
    }
  }
}

File: src-tauri/src/api/catalog.rs
use crate::api::config::Config;
use crate::api::utils::generate_fields::generate_fields;
use crate::api::utils::generate_request::generate_request;
use crate::api::utils::generate_url::{generate_url, generate_keyed_params};
use crate::api::utils::generate_user_agent::generate_user_agent;

#[derive(Clone)]
pub struct CatalogManager {
    config: Config,
    client: reqwest::Client,
}

impl CatalogManager {
    pub fn new(config: &Config, client: reqwest::Client) -> Self {
        Self {
            config: config.clone(),
            client,
        }
    }
    // Поиск по всем манга подсайтам из списка
    pub async fn search_manga_objects(&self, query: String, site_ids: Option<Vec<&str>>) -> Result<String, reqwest::Error> {
        let site_ids = site_ids.unwrap_or(vec!["1", "2", "3", "4"]);
        let site_ids_parts = generate_keyed_params("site_ids", &site_ids);
        let mut url = generate_url(
            &self.config.base_url,
            "/manga",
            &[],
            Some(vec![
                ("q", &query),
            ]),
            Some(vec![
                "rate_avg", "rate", "releaseDate"
            ]),
        );
        url = format!("{}{}", url, site_ids_parts.join("&"));

        generate_request(&self.client, &url, "manga", &generate_user_agent()).await
    }

    // Поиск по всем аниме
    pub async fn search_anime_objects(&self, query: String) -> Result<String, reqwest::Error> {
        let url = generate_url(
            &self.config.base_url,
            "/anime",
            &[],
            Some(vec![
                ("q", &query),
            ]),
            Some(vec![
                "rate_avg", "rate", "releaseDate"
            ]),
        );

        generate_request(&self.client, &url, "anime", &generate_user_agent()).await
    }

    // Поиск по переводчикам и тд
    pub async fn search_teams(&self, query: String) -> Result<String, reqwest::Error> {
        let url = generate_url(
            &self.config.base_url,
            "/teams",
            &[],
            Some(vec![
                ("q", &query),
            ]),
            None,
        );

        generate_request(&self.client, &url, "anime", &generate_user_agent()).await
    }

    // Поиск по персонажам
    pub async fn search_character(&self, query: String) -> Result<String, reqwest::Error> {
        let url = generate_url(
            &self.config.base_url,
            "/character",
            &[],
            Some(vec![
                ("q", &query),
            ]),
            None,
        );

        generate_request(&self.client, &url, "anime", &generate_user_agent()).await
    }

    // Поиск по людям (автора)
    pub async fn search_people(&self, query: String) -> Result<String, reqwest::Error> {
        let url = generate_url(
            &self.config.base_url,
            "/people",
            &[],
            Some(vec![
                ("q", &query),
            ]),
            None,
        );

        generate_request(&self.client, &url, "anime", &generate_user_agent()).await
    }

    // Поиск по франшизам
    pub async fn search_franchise(&self, query: String) -> Result<String, reqwest::Error> {
        let url = generate_url(
            &self.config.base_url,
            "/franchise",
            &[],
            Some(vec![
                ("q", &query),
            ]),
            None,
        );

        generate_request(&self.client, &url, "anime", &generate_user_agent()).await
    }

    // Поиск по публицистам (типо студия выпуска)
    pub async fn search_publisher(&self, query: String) -> Result<String, reqwest::Error> {
        let url = generate_url(
            &self.config.base_url,
            "/publisher",
            &[],
            Some(vec![
                ("q", &query),
            ]),
            None,
        );

        generate_request(&self.client, &url, "anime", &generate_user_agent()).await
    }

    // Поиск по пользователям
    pub async fn search_user(&self, query: String, sort_by: Option<&str>, sort_type: Option<&str>) -> Result<String, reqwest::Error> {
        let sort_by = sort_by.unwrap_or("name");
        let sort_type = sort_type.unwrap_or("desc");

        let url = generate_url(
            &self.config.base_url,
            "/user",
            &[],
            Some(vec![
                ("q", &query),
                ("sort_by", sort_by),
                ("sort_type", sort_type),
            ]),
            None,
        );

        generate_request(&self.client, &url, "anime", &generate_user_agent()).await
    }

}
